src/adoption/adoption-token.ts

```diff
@@ -0,0 +1,218 @@
+import type { Knex } from 'knex';
+import { randomUUID } from 'crypto';
+import { scopedLogger } from '../lib/logger.js';
+import { client } from '../lib/sql/client.js';
+import { Probe } from '../probe/types.js';
+import { ServerSocket, adoptedProbes } from '../lib/ws/server.js';
+import { AdoptedProbes } from '../lib/override/adopted-probes.js';
+
+const USERS_TABLE = 'directus_users';
+const PROBES_TABLE = 'gp_probes';
+const NOTIFICATIONS_TABLE = 'directus_notifications';
+
+const logger = scopedLogger('adoption-token');
+
+type User = {
+	id: string;
+	adoption_token: string;
+};
+
+type DProbe = {
+	id: string;
+	name: string | null;
+	ip: string;
+	userId: string | null;
+}
+
+export class AdoptionToken {
+	private tokensToUsers = new Map<string, User>();
+	private timer: NodeJS.Timeout | undefined;
+
+	constructor (
+		private readonly sql: Knex,
+		private readonly adoptedProbes: AdoptedProbes,
+	) {}
+
+	scheduleSync () {
+		clearTimeout(this.timer);
+
+		this.timer = setTimeout(() => {
+			this.syncTokens()
+				.finally(() => this.scheduleSync())
+				.catch(error => logger.error('Error in AdoptionToken.syncTokens()', error));
+		}, 60_000).unref();
+	}
+
+	unscheduleSync () {
+		clearTimeout(this.timer);
+	}
+
+	async syncTokens () {
+		const users = await this.fetchUsers();
+		this.tokensToUsers = new Map(users.map(user => ([ user.adoption_token, user ])));
+	}
+
+	private async fetchSpecificUser (token: string) {
+		const users = await this.fetchUsers({ adoption_token: token });
+
+		if (users.length === 0) {
+			return undefined;
+		}
+
+		const user = users[0]!;
+
+		this.tokensToUsers.set(user.adoption_token, user);
+		return user;
+	}
+
+	private async fetchUsers (filter: Record<string, unknown> = {}) {
+		const users = await this.sql(USERS_TABLE)
+			.where(filter)
+			.select<User[]>([ 'id', 'adoption_token' ]);
+		return users;
+	}
+
+	async validate (socket: ServerSocket) {
+		const tokenValue = socket.handshake.query['adoptionToken'];
+		const token = !tokenValue ? null : String(tokenValue);
+		const probe = socket.data.probe;
+		const isAdopted = !!adoptedProbes.getByIp(probe.ipAddress);
+
+		if (!token) {
+			!isAdopted && socket.emit('api:connect:adoption', { message: 'You can register this probe at https://dash.globalping.io to earn extra measurement credits.' });
+			return;
+		}
+
+		const message = await this.validateToken(token, probe);
+
+		if (message) {
+			socket.emit('api:connect:adoption', { message });
+		}
+	}
+
+	async validateToken (token: string, probe: Probe) {
+		let user = this.tokensToUsers.get(token);
+
+		if (!user) {
+			user = await this.fetchSpecificUser(token);
+		}
+
+		if (!user) {
+			logger.info('User not found for the provided adoption token.', { token });
+			return `User not found for the provided adoption token: ${token}.`;
```

Comment:
This seems like it should be a warning on the probe, so I guess send a `level` as well?